File: 0121_best_time_to_buy_and_sell_stock/solution.py
from typing import List

# runtime: 65.89%, memory: 94.84%
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        max_profit: int = 0
        buy, sell = 0, 1
        length: int = len(prices)

        while sell < length:
            if prices[buy] < prices[sell]:
                max_profit = max(max_profit, prices[sell] - prices[buy])
            else:
                buy = sell
            sell += 1

        return max_profit

        

File: 0121_best_time_to_buy_and_sell_stock/Solution.java
import Math;

// runtime: 74.94%, memory: 74.00%
class Solution {
    public int maxProfit(int[] prices) {
        int maxProfit = 0;
        int buy = 0;
        int sell = 1;
        int length = prices.length;

        while (sell < length) {
            if (prices[buy] < prices[sell]) {
                maxProfit = Math.max(maxProfit, prices[sell] - prices[buy]);
            } else {
                buy = sell;
            }
            sell++;
        }

        return maxProfit;
    }
}